### PR TITLE
DataMovement Resource Providers

### PR DESCRIPTION
**Pull request by jaschrep-msft**
Creates static methods accepting checkpointer data which return "resource providers." Resource providers will recreate the resource used in a transfer with the same configurations the transfer was started with.

### CODE REVIEW

File: sdk/storage/Azure.Storage.DataMovement.Blobs/src/AzureBlobStorageResourceProvider.cs [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+                throw BadResourceTypeException(resourceType);
+            }
+
+            Properties = properties;
+            MakesSource = asSource;
+            ResourceType = resourceType;
+        }
+
+        /// <summary>
+        ///
```
Comment (jalauzon-msft) [inline sdk/storage/Azure.Storage.DataMovement.Blobs/src/AzureBlobStorageResourceProvider.cs:38]: Doc comments in here?

File: sdk/storage/Azure.Storage.DataMovement.Blobs/src/AzureBlobStorageResources.cs [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+        private static (ResourceType SourceType, ResourceType DestinationType) GetTypes(DataTransferProperties info)
+        {
+            Argument.AssertNotNull(info, nameof(info));
+
+            ResourceType GetType(string scheme)
+            {
+                return scheme switch
+                {
+                    // TODO figure out actual strings
```
Comment (jalauzon-msft) [inline sdk/storage/Azure.Storage.DataMovement.Blobs/src/AzureBlobStorageResources.cs:70]: They are `BlockBlob`, `PageBlob`, `AppendBlob`, and `LocalFile` (for the local one). They should match the new `ResourceId` property on the resource classes.

Comment (jaschrep-msft) [reply sdk/storage/Azure.Storage.DataMovement.Blobs/src/AzureBlobStorageResources.cs:70]: hmm I had forgotten about this bit. It seems I'm only testing the instanced resource provider and not the static method invocation. That needs fixing.

File: sdk/storage/Azure.Storage.DataMovement.Blobs/api/Azure.Storage.DataMovement.Blobs.net6.0.cs [hunk 3]
```diff
@@ -42,6 +42,18 @@ public partial class AppendBlobStorageResourceOptions : Azure.Storage.DataMoveme
         public Azure.Storage.Blobs.Models.AppendBlobRequestConditions DestinationConditions { get { throw null; } set { } }
         public Azure.Storage.Blobs.Models.AppendBlobRequestConditions SourceConditions { get { throw null; } set { } }
     }
+    public partial class AzureBlobStorageResourceProvider
```
Comment (amnguye) [inline sdk/storage/Azure.Storage.DataMovement.Blobs/api/Azure.Storage.DataMovement.Blobs.net6.0.cs:45]: Maybe thinking of dropping the word Azure and just having `BlobStorageResourceProvider`.

Same with the below with `BlobStorageResources`

**Review by jalauzon-msft — APPROVED** [review]

**Pull request merged**